### ISSUE TITLE
Can io_uring_prep_read the tun fd with Multishot?

### ISSUE THREAD

**Issue by pyhd** [axboe/liburing#851]
If multishot can work on the socket, could it be possible to apply this on the read operation?

**Comment by axboe** [comment]
The io_uring read operations currently don't support multishot, but there's no reason why it could not be provided the target is pollable (like tun would be, or a socket).

**Comment by illiliti** [comment]
Why closed? If it's feature request, why not leave open? IMO multishot read would be great addition since with it you can read fd directly without doing needless roundtrips via multishot poll. Of course it should work only with pollable descriptors like tun fd, signalfd, timerfd and so on.

**Comment by axboe** [comment]
It would not be hard to support. Here's a 5 min hack that does just that:

https://git.kernel.dk/cgit/linux/log/?h=io_uring-mshot-read

you can use it from an app by doing:

```
io_uring_prep_read(sqe, fd, NULL, buf_size, 0);
sqe->opcode = IORING_OP_READ_MULTISHOT;
sqe->buf_group = buffer_group_id;
sqe->flags |= IOSQE_BUFFER_SELECT;
```

and a provided buffer group must be setup as well, as multishot can only work with a provided buffer group rather than passing in a single buffer.

**Comment by axboe** [comment]
Simple test case:

```
/* SPDX-License-Identifier: MIT */
/*
 * Description: Test READ_MULTISHOT with ring mapped provided buffers
 *
 */
#include <errno.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>

#include "liburing.h"
#include "helpers.h"

#define BUF_SIZE	32
#define NR_BUFS		64
#define BUF_BGID	1

#define BR_MASK		(NR_BUFS - 1)

static int no_buf_ring;

static int test(void)
{
	struct io_uring_sqe *sqe;
	struct io_uring_cqe *cqe;
	struct io_uring ring;
	struct io_uring_buf_ring *br;
	int ret, fds[2], i;
	char *buf;
	void *ptr;

	ret = io_uring_queue_init(NR_BUFS, &ring, 0);
	if (ret) {
		fprintf(stderr, "ring setup failed: %d\n", ret);
		return 1;
	}

	if (pipe(fds) < 0) {
		perror("pipe");
		return 1;
	}

	if (posix_memalign((void **) &buf, 4096, NR_BUFS * BUF_SIZE))
		return 1;

	br = io_uring_setup_buf_ring(&ring, NR_BUFS, BUF_BGID, 0, &ret);
	if (!br) {
		if (ret == -EINVAL) {
			no_buf_ring = 1;
			return 0;
		}
		fprintf(stderr, "Buffer ring register failed %d\n", ret);
		return 1;
	}

	ptr = buf;
	for (i = 0; i < NR_BUFS; i++) {
		io_uring_buf_ring_add(br, ptr, BUF_SIZE, i + 1, BR_MASK, i);
		ptr += BUF_SIZE;
	}
	io_uring_buf_ring_advance(br, NR_BUFS);

	sqe = io_uring_get_sqe(&ring);
	io_uring_prep_read(sqe, fds[0], NULL, BUF_SIZE, 0);
	sqe->opcode = IORING_OP_READ_MULTISHOT;
	sqe->buf_group = BUF_BGID;
	sqe->flags |= IOSQE_BUFFER_SELECT;
	sqe->user_data = 1;

	ret = io_uring_submit(&ring);
	if (ret != 1) {
		fprintf(stderr, "submit: %d\n", ret);
		return 1;
	}

	for (i = 0; i < NR_BUFS + 1; i++) {
		char tmp[32];

		ret = write(fds[1], tmp, strlen(tmp));
		if (ret != strlen(tmp)) {
			printf("write ret %d\n", ret);
			return 1;
		}
		usleep(1000);
	}

	for (i = 0; i < NR_BUFS + 1; i++) {
		int bid;

		ret = io_uring_wait_cqe(&ring, &cqe);
		if (ret) {
			fprintf(stderr, "wait cqe failed %d\n", ret);
			return 1;
		}
		printf("got cqe res %d, flags %x\n", cqe->res, cqe->flags);
		if (!(cqe->flags & IORING_CQE_F_BUFFER)) {
			fprintf(stderr, "no buffer selected\n");
			return 1;
		}
		bid = cqe->flags >> IORING_CQE_BUFFER_SHIFT;
		printf("  bid=%d\n", bid);
		if (!(cqe->flags & IORING_CQE_F_MORE)) {
			fprintf(stderr, "no more cqes\n");
			break;
		}
	
		io_uring_cqe_seen(&ring, cqe);
	}

	return 0;
}

int main(int argc, char *argv[])
{
	int ret;

	if (argc > 1)
		return T_EXIT_SKIP;

	ret = test();
	if (ret) {
		fprintf(stderr, "dio test failed\n");
		return T_EXIT_FAIL;
	}
	if (no_buf_ring)
		return T_EXIT_SKIP;

	return T_EXIT_PASS;
}
```

**Comment by illiliti** [comment]
That was quick, thanks! Unfortunately I can't test those patches since they are based on latest linux which zfs does not support yet, but from what I see this is exactly what I'm personally looking for. Is there a plan to upstream them or it's just a PoC?

**Comment by illiliti** [comment]
Excellent! Looking forward to see it upstream!

**Comment by illiliti** [comment]
Forgot to ask: will there be a POLL_FIRST flag to not reserve buffer immediately but only when data arrive? Or it is not needed?

**Comment by axboe** [comment]
We could add that. But if you use ring mapped provided buffers, then it'll recycle just fine and at minimum cost. I think the main win with POLL_FIRST would be that you avoid trying a read that you know will fail, only to retry it when the poll callback triggers.

However, for multishot, it's less important as this would only be a concern for the very first attempt. Any subsequent attempt should only trigger when there is data to read, and POLL_FIRST wouldn't apply there in any case.

Hence I don't think it's worth complicating it with a POLL_FIRST flag.

**Comment by pyhd** [comment]
One more question: Is it feasible to combine `register_eventfd `+ `IORING_FEAT_FAST_POLL` + `multishot read/receive`?  In other words, can `eventfd` triggers only after `multishot` reads/receives all pendings? So in the event callback, userspace can re-arm `multishot` with `IORING_FEAT_FAST_POLL`.

**Comment by axboe** [comment]
Not sure I follow. IORING_FEAT_FAST_POLL is just an internal feature, that tells you that operating on pollable files is efficient. It predates multishot <anything> by a lot, and is also what drives that. You will not find a kernel that supports multishot and doesn't have the fast poll feature.

For the latter part, no it's not possible to have eventfd only trigger if a multishot request fails. And that is generally the only case where it will go away, it's not like you arm it for 10 receives and then it terminates.

**Comment by pyhd** [comment]
My idea is to trigger a POLLIN eventfd after reading a bulk of packets like 64K Bytes. In essence, just to replace "the epoll of IOs" with "the epoll of CQEs". Hence my concern is whether the eventfd would trigger upon : (A) any CQE or (B) CQE without `ioring_cqe_f_more` flag.

**Comment by axboe** [comment]
It'd trigger on any CQE, only ones that don't are:

1) Ones that have SKIP_SUCCESS set
2) Ones that completed inline AND IORING_REGISTER_EVENTFD_ASYNC was used to register the eventfd

For what you would like, we'd need a new way to register the eventfd.

How are you planning on managing this? As mentioned, the multishot read will keep posting CQEs with CQE_MORE set until something unexpected happens. Granted you could make it terminate if you manage your provided buffer pool, eg it's always replenished with N buffers in it, and hence you know that N+1 read will result in a failed request with -ENOBUFS set. It all just seems a bit weird and maybe just too specific to be generically usable. I'm happy to be convinced otherwise, these are just my thoughts on your proposal.

**Comment by pyhd** [comment]
Thank you for your clarification.

I think I should demonstrate this model further, because people might wonder the difference and necessity.

In a traditional Epoll (IO based) model: epoll_wait -> read(or receive) -> data process -> send(or write);
In a new Epoll (CQE based) model: epoll_wait -> data process -> batch send&read (or write&receive).

Here I just borrow the idea from pipelining: the data, ready in the N round, was submitted in the previous N-1 round operation. Therefore this new model reduces the syscalls from 3 to 2, and at the same time, no need to wait for the read(or receive) operation in the traditional model. Once kernel wakes up userspace, the data is ready to hand.

If I'm not mistaken, your concern is the eventfd should trigger upon , not only CQE without cqe_f_more flag (within the boundary), but also -ENOBUFS. Surely it is my fault to not consider about -ENOBUFS. 

As for such kind of latency and syscall optimization, I cannot say to which extend it could achieve in terms of performance. Probably large or trivial, but it is still not a bad idea to fix such mismatch between eventfd and multishot, even at the cost of new helper or flag.

**Comment by axboe** [comment]
I think part of the problem here is that you're still using epoll on the io_uring fd to collect events. If you were using cqring_wait directly, then you could very easily just say "wait for X time, or until Y events are available". This isn't something that's usable from the poll/epoll side, obviously. And another important note here is that eventfd usage isn't free either. I'd wager that you're better off doing 3 syscalls rather than 2, where the 2 come with needing to use epoll and eventfd as well.

Just to clarify -ENOBUFS - what I mean here is that you'll always get a CQE posted, regardless of whether SKIP_SUCCESS is set or not, as it's an error condition. As such, a posted CQE like that would ALWAYS trigger the eventfd.

It's easy to do this via waiting on events, as that context is private per wait. It's a bit harder to express with a registered entity like eventfd. You could have a way to say "trigger eventfd whenever N successful events have happened" and include things like "an errored CQE would also trigger it".

**Comment by pyhd** [comment]
Your words help me a lot, really appreciate that. Now I find the beauty of io_uring to replace `epoll_wait` with `wait_cqe` in a loop. Thus in my case, the syscall(s) can be reduced from 3 to 1, right?

**Comment by pyhd** [comment]
@axboe Test on kernel 6.7, two questions:
1. If pass non zero parameter `nbytes` to `io_uring_prep_read_multishot`, `cqe->res` returns `-EINVAL` and `cqe->flags` doesn't set `IORING_CQE_F_BUFFER`.
2. `cqe->flags` doesn't set `IORING_CQE_F_MORE`, Is it a design on purpose? So multishot will never stop until some error or cancellation. In order to fetch multiple reads within a short interval, is it better to use `io_uring_wait_cqes`?

**Comment by axboe** [comment]
Yes, you should not set len, it's dictated by the size of the buffers in the buffer group ID given. It does set IORING_CQE_F_MORE in case of error, because there will be no more triggers of it in that case.

> In order to fetch multiple reads within a short interval, is it better to use io_uring_wait_cqes?

That one I don't understand. It's always better to wait for multiple events in terms of amortizing the cost of doing that, but that's not really related to multishot, that's a general principle.

**Comment by axboe** [comment]
I'm going to close this one as it was for the original introduction of read multishot. If there are questions related to it, just open another issue.
